cladecombiner/aggregator.py

```diff
@@ -0,0 +1,228 @@
+            raise RuntimeError(
+                "Mismatch between aggregated taxa and input taxa. Input taxa are: "
+                + str(input_taxa)
+                + " but aggregated taxa are "
+                + str(taxon_map.keys())
+            )
+        tab = table(taxon_map)
+        if not all(v == 1 for v in tab.values()):
+            raise RuntimeError(
+                "Found following taxa mapped more than once: "
+                + str([k for k, v in tab.items() if v > 1])
+            )
+
+    def __init__(
+        self, input_taxa: Iterable[Taxon], taxon_map: dict[Taxon, Taxon]
+    ):
+        self._validate(input_taxa, taxon_map)
+        super().__init__(taxon_map)
+
+    def to_str(self):
+        """
+        Get str : str map of taxa names
+        """
+        return {k.name: v.name for k, v in self.items()}
+
+
+class Aggregator(ABC):
+    """
+    Aggregators return Aggregations, maps of input_taxon : aggregated_taxon
+    """
+
+    @abstractmethod
+    def aggregate(self, input_taxa: Iterable[Taxon]) -> Aggregation:
+        pass
+
+
+class ArbitraryAggregator(Aggregator):
+    """
+    Aggregation via a user-provided dictionary.
+    """
+
+    def __init__(
+        self,
+        map: dict[Taxon, Taxon],
+    ):
+        """
+        FixedAggregator constructor.
+
+        Parameters
+        ----------
+        map : dict[Taxon, Taxon]
+            Dictionary mapping the input taxa to their aggregated taxa.
+        """
+        self.map = map
+
+    def aggregate(self, input_taxa: Iterable[Taxon]) -> Aggregation:
+        return Aggregation(
+            input_taxa, {taxon: self.map[taxon] for taxon in input_taxa}
+        )
+
+
+class BasicPhylogeneticAggregator(Aggregator):
+    """
+    An aggregator which maps a set of input taxa to a fixed set of aggregation targets using a tree.
+    """
+
+    def __init__(
+        self,
+        targets: Iterable[Taxon],
+        taxonomy_scheme: PhylogeneticTaxonomyScheme,
+        sort_clades: bool = True,
+        unmapped_are_other: bool = True,
+        warn: bool = True,
+    ):
+        """
+        BasicPhylogeneticAggregator constructor.
+
+        Parameters
+        ----------
+        targets : Iterable[Taxon]
+            The taxa into which we wish to aggregate the input taxa.
+
+        taxonomy_scheme : PhylogeneticTaxonomyScheme
+            The tree which we use to do the mapping.
+
+        sort_clades : bool
+            If False, mapping is done using the taxa as ordered in `targets`.
+            If True, `targets` are taxonomically sorted so that so that larger
+            `targets` do not override smaller ones. For example, if BA.2 and
+            BA.2.86 are both aggregation targets, sort_clades = True would handle
+            BA.2.86 first, such that JN.1 would map to BA.2.86, while BG.1 would
+            map to BA.2. If BA.2 is processed first, both will map to it.
+
+        unmapped_are_other : bool
```

Comment:
[low priority] My understanding is that better practice would be to make this an argument like `off_target=`, which can take values like `"other"`, or `"self"` or something. We'd need some better nomenclature.

But this gives you a little future-proofing, so that if there are eventually more than 2 ways to deal with off-target taxa, you don't need to break the interface

Reply:
That's fair. No reason to back into a corner for lack of imagination, especially when the alternative is at least as syntactically clear.